azure/scope/workload_identity.go

```diff
@@ -0,0 +1,155 @@
+|                           | application or user-assigned managed identity.        |
+|AZURE_TENANT_ID            | The tenant ID of the Azure subscription.              |
+|AZURE_FEDERATED_TOKEN_FILE | The path of the projected service account token file. |
+|-----------------------------------------------------------------------------------|
+
+In addition to the service account, it also projects a signed service account token to the
+workload's volume(in this case the capz pod). The volume name is `azure-identity-token`
+which is mounted at path `/var/run/secrets/azure/tokens/azure-identity-token` to the pod.
+
+*/
+
+const (
+	// AzureFedratedTokenFileENVKey is the env key for AZURE_FEDERATED_TOKEN_FILE.
+	AzureFedratedTokenFileENVKey = "AZURE_FEDERATED_TOKEN_FILE"
+	// AzureClientIDENVKey is the env key for AZURE_CLIENT_ID.
+	AzureClientIDENVKey = "AZURE_CLIENT_ID"
+	// AzureTenantIDENVKey is the env key for AZURE_TENANT_ID.
+	AzureTenantIDENVKey = "AZURE_TENANT_ID"
+)
+
+type workloadIdentityCredential struct {
+	assertion string
+	file      string
+	cred      *azidentity.ClientAssertionCredential
+	lastRead  time.Time
+}
+
+// WorkloadIdentityCredentialOptions contains the configurable options for azwi.
+type WorkloadIdentityCredentialOptions struct {
+	ClientID      string
+	TenantID      string
+	TokenFilePath string
+	azcore.ClientOptions
+}
+
+// NewWorkloadIdentityCredentialOptions returns an empty instance of WorkloadIdentityCredentialOptions.
+func NewWorkloadIdentityCredentialOptions() *WorkloadIdentityCredentialOptions {
+	return &WorkloadIdentityCredentialOptions{}
+}
+
+// WithClientID sets client ID to WorkloadIdentityCredentialOptions.
+func (w *WorkloadIdentityCredentialOptions) WithClientID(clientID string) *WorkloadIdentityCredentialOptions {
+	w.ClientID = clientID
+	return w
+}
+
+// WithTenantID sets tenant ID to WorkloadIdentityCredentialOptions.
+func (w *WorkloadIdentityCredentialOptions) WithTenantID(tenantID string) *WorkloadIdentityCredentialOptions {
+	w.TenantID = tenantID
+	return w
+}
+
+// GetProjectedTokenPath return projected token file path from the env variable.
+func GetProjectedTokenPath() (string, error) {
+	tokenPath := os.Getenv(AzureFedratedTokenFileENVKey)
+	if strings.TrimSpace(tokenPath) == "" {
+		return "", errors.New("projected token path not injected")
+	}
+	return tokenPath, nil
+}
+
+// WithDefaults sets token file path. It also sets the client tenant ID from injected env in
+// case empty values are passed.
+func (w *WorkloadIdentityCredentialOptions) WithDefaults() (*WorkloadIdentityCredentialOptions, error) {
+	tokenFilePath, err := GetProjectedTokenPath()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get token file path for identity")
+	}
+	w.TokenFilePath = tokenFilePath
+
+	// Fallback to using client ID from env variable if not set.
+	if strings.TrimSpace(w.ClientID) == "" {
+		w.ClientID = os.Getenv(AzureClientIDENVKey)
```

Comment:
Is CAPZ going to always have the user provide the clientID as part of the custom resource spec? If that's the case when will this value ever be not empty?